s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/S3SourceTask.java

```diff
@@ -144,24 +144,35 @@ private void prepareReaderFromOffsetStorageReader() {
 
     @Override
     public List<SourceRecord> poll() throws InterruptedException {
-        LOGGER.info("Polling again");
+        LOGGER.info("Polling for new records...");
         synchronized (pollLock) {
             final List<SourceRecord> results = new ArrayList<>(s3SourceConfig.getInt(MAX_POLL_RECORDS));
 
             if (connectorStopped.get()) {
+                LOGGER.info("Connector has been stopped. Returning empty result list.");
                 return results;
             }
 
             while (!connectorStopped.get()) {
                 try {
-                    LOGGER.info("Number of records sent {}", extractSourceRecords(results).size());
+                    extractSourceRecords(results);
+                    LOGGER.info("Number of records extracted and sent: {}", results.size());
                     return results;
-                } catch (AmazonS3Exception | DataException exception) {
-                    if (handleException(exception)) {
+                } catch (AmazonS3Exception exception) {
+                    if (exception.isRetryable()) {
+                        LOGGER.warn("Retryable error encountered during polling. Waiting before retrying...",
+                                exception);
+                        pollLock.wait(ERROR_BACKOFF);
+
+                        prepareReaderFromOffsetStorageReader();
+                    } else {
+                        LOGGER.warn("Non-retryable AmazonS3Exception occurred. Stopping polling.", exception);
                         return null; // NOPMD
                     }
+                } catch (DataException exception) {
+                    LOGGER.warn("DataException occurred during polling. No retries will be attempted.", exception);
```

Comment:
Is this correct?  The way I read it, it will immediately retry the while loop without waiting (both before and after the change).

Reply:
Yes, we are ignoring the exception here with a warning. We shall revisit this section in handle malformed records ticket I believe.